### PR TITLE
feat: add basic payload generator

### PR DESCRIPTION
**Pull request by mattsse**
~~Builds off of #2157~~

Adds an implementation of `PayloadJobGenerator` that uses the txpool for filling a block

This `PayloadJob` fills block on an interval basis, if the next block is better, it replaces the old block.
for the basic job, better just means higher total fees.

### CODE REVIEW

File: crates/payload/basic/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+            }));
+            this.pending_block = Some(PendingPayload { _cancel, payload: rx });
+        }
+
+        // poll the pending block
+        if let Some(mut fut) = this.pending_block.take() {
+            match fut.poll_unpin(cx) {
+                Poll::Ready(Ok(payload)) => {
+                    this.interval.reset();
+                    // TODO check if payload is better
```
Comment (Rjected) [inline crates/payload/basic/src/lib.rs:232]: would we just compare the `BuiltPayload` fees here? Is there anything else involved with checking that the payload is better?

**Review by Rjected — COMMENTED** [review]
just had a question here about best payloads. Did a pass and I really like the structure so far! almost missed `crates/payload/basic/src/lib.rs` because of github folding though

File: crates/payload/basic/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+            // append transaction to the list of executed transactions
+            executed_txs.push(tx);
+
+            // append gas used
+            cumulative_gas_used += res.result.gas_used();
+        }
+
+        // Process withdrawals
+        if chain_spec.fork(Hardfork::Shanghai).active_at_timestamp(attributes.timestamp) {
```
Comment (rkrasiuk) [inline crates/payload/basic/src/lib.rs:397]: `initialized_cfg` already contains revm spec that can be used to check if it's shanghai or not. this allows making `PayloadConfig` slimmer

Comment (mattsse) [reply crates/payload/basic/src/lib.rs:397]: TIL

File: crates/payload/basic/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            // convert tx to a signed transaction
+            let tx = tx.to_recovered_transaction();
+            let env = Env {
+                cfg: initialized_cfg.clone(),
+                block: initialized_block_env.clone(),
+                tx: tx_env_with_recovered(&tx),
+            };
+
+            let mut evm = revm::EVM::with_env(env);
+            evm.database(&mut db);
```
Comment (rkrasiuk) [inline crates/payload/basic/src/lib.rs:384]: add a comment that you will need to reinitialize the database with current post state

File: crates/payload/basic/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+            // TODO skip invalid transactions
+            let res = evm.transact().map_err(PayloadBuilderError::EvmExecutionError)?;
+
+            // append transaction to the list of executed transactions
+            executed_txs.push(tx);
+
+            // append gas used
+            cumulative_gas_used += res.result.gas_used();
```
Comment (rkrasiuk) [inline crates/payload/basic/src/lib.rs:393]: you can already call `commit_changes` here and apply them to `post_state`

File: crates/payload/basic/src/lib.rs (hunk 1 above)
Comment (mattsse) [reply crates/payload/basic/src/lib.rs]: by default no, it's just:

https://github.com/ethereum/go-ethereum/blob/4a9fa31450d3cdcea84735b68cd5a0a8450473f8/miner/payload_building.go#L96-L99

File: crates/payload/basic/src/lib.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/payload/basic/src/lib.rs:472]: current problem is that it's executing **each** transaction on top of the state of the previous block (it doesn't account for just executed transactions)

Comment (mattsse) [reply crates/payload/basic/src/lib.rs:472]: I see, a lot of block traces then also suffer from this, 

what's the right fix for this?

Comment (rkrasiuk) [reply crates/payload/basic/src/lib.rs:472]: ah, i see that commit_changes impl was changed, seems like some time ago, to write to cache db as well

**Review by rkrasiuk — APPROVED** [review]
lgtm overall. one thing to think about is that we don't need to construct a full block (perform expensive calculations like header roots computation) on each iteration. it'd be more performant to return transaction ordering + post state + gas used and construct full block only when we need to submit it

**Comment by gakonst** [comment]
I think Roman means that there's a bunch of expensive ops per block like generating the transaction/state roots etc. and you could probably avoid the need for calculating them on every built block.

File: crates/revm/revm-primitives/src/config.rs [hunk 5]
```diff
@@ -2,6 +2,21 @@
...
+/// Returns the spec id at the given timestamp.
+///
+/// Note: This is only intended to be used after the merge, when hardforks are activated by
+/// timestamp.
+pub fn revm_spec_by_timestamp_after_merge(
+    chain_spec: &ChainSpec,
+    timestamp: u64,
+) -> revm::primitives::SpecId {
+    if chain_spec.is_fork_active_at_timestamp(Hardfork::Shanghai, timestamp) {
+        revm::primitives::SHANGHAI
```
Comment (gakonst) [inline crates/revm/revm-primitives/src/config.rs:14]: Should we have something like `chain_spec.latest_fork_at(timestamp)`?

File: crates/payload/basic/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+            timestamp: U256::from(attributes.timestamp),
+            difficulty: U256::ZERO,
+            prevrandao: Some(attributes.prev_randao),
+            gas_limit: U256::from(parent_block.gas_limit),
+            // calculate basefee based on parent block's gas usage
+            basefee: U256::from(calculate_next_block_base_fee(
+                parent_block.gas_used,
+                parent_block.gas_limit,
+                parent_block.base_fee_per_gas.unwrap_or_default(),
+            )),
```
Comment (gakonst) [inline crates/payload/basic/src/lib.rs:144]: need a `BlockEnv::new(&block, attrs)`

File: crates/payload/basic/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+            best_payload: None,
+            pending_block: None,
+            payload_task_guard: self.payload_task_guard.clone(),
+        })
+    }
+}
+
+/// Restricts how many generator tasks can be executed at once.
+#[derive(Clone)]
+struct PayloadTaskGuard(Arc<Semaphore>);
```
Comment (gakonst) [inline crates/payload/basic/src/lib.rs:177]: really nice pattern, in general will be really interesting to hear you explain / talk payload building. also maybe we should combine payload and miner into some top-level crate? or put them in a shared subdir?

File: crates/payload/basic/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+    fn is_better(&self, new_payload: &BuiltPayload) -> bool {
+        if let Some(best_payload) = &self.best_payload {
+            new_payload.fees() > best_payload.fees()
+        } else {
+            true
+        }
+    }
+}
+
+impl<Client, Pool, Tasks> Stream for BasicPayloadJob<Client, Pool, Tasks>
```
Comment (gakonst) [inline crates/payload/basic/src/lib.rs:273]: I like this pattern, each job is responsible for everything on its own, vs having some master task that handles things itself, the master task just drives the jobs a nd the complexity is encapsulated there

**Review by gakonst — APPROVED** [review]
supportive and approving to get this merged and unblock us from building at the tip as a proposer, but would love some more tests here as we're prob bound to encoutner edge cases

**Comment by mattsse** [comment]
cool, I have a bunch of things I need to improve, will open issues for them shortly.
